lib/streamlit/server/UploadFileRequestHandler.py

```diff
@@ -0,0 +1,128 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018-2020 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Any
+
+import tornado.web
+import tornado.httputil
+
+from streamlit.UploadedFileManager import UploadedFile
+from streamlit.logger import get_logger
+from streamlit.server import routes
+
+LOGGER = get_logger(__name__)
+
+
+class UploadFileRequestHandler(tornado.web.RequestHandler):
+    """
+    Implements the PUT /upload_file endpoint.
+    """
+
+    def initialize(self, file_mgr):
+        """
+        Parameters
+        ----------
+        file_mgr : UploadedFileManager
+            The server's singleton UploadedFileManager. All file uploads
+            go here.
+
+        """
+        self._file_mgr = file_mgr
+
+    def set_default_headers(self):
+        if routes.allow_cross_origin_requests():
+            self.set_header("Access-Control-Allow-Origin", "*")
```

Comment:
is this secure?

Reply:
Yes - this is the same header we set on all HTTP handlers when the `allow_cross_origin_requests()` function is True (which is when the user is in development mode, or has explicitly requested it via config). 

(By default, the "Allow *" header will not be added; it's just in these two cases; the `allow_cross_origin_requests` function has more details.)